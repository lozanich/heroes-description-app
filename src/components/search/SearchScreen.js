import React, { useMemo } from "react";
import queryString from "query-string";
import { useLocation } from "react-router-dom";
// import { heroes } from "../../data/heroes";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heroes/HeroCard";
import { getHeroesByName } from "../../selectors/getHeroesByName";

export const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);
  const { values, handleInputChange } = useForm({
    nameHero: q,
  });

  const { nameHero } = values;
  //   console.log(nameHero);

  let heroesFilter = useMemo(() => getHeroesByName(q), [q]);

  //   const heroesFilter = getHeroesByName(nameHero);

  const handleSearch = (e) => {
    e.preventDefault();
    if (nameHero.trim().length < 1) {
      return;
    }
    history.push(`?q=${nameHero}`);
    // reset();
  };

  return (
    <div>
      <h1>Search Screen</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />

          <form onSubmit={handleSearch} className="form-group">
            <input
              className="form-control"
              type="text"
              onChange={handleInputChange}
              name="nameHero"
              autoComplete="off"
              value={nameHero}
              placeholder="Find your Hero"
            />

            <button
              type="submit"
              className="btn m-1 btn-block btn-outline-primary"
            >
              Search....
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {q === "" && <div className="alert alert-info">Busca un Heroe</div>}

          {q !== "" && heroesFilter.length === 0 && (
            <div className="alert alert-danger">
              No se encontraron resultados
            </div>
          )}

          {heroesFilter.map((hero) => {
            return <HeroCard key={hero.id} hero={hero} />;
          })}
        </div>
      </div>
    </div>
  );
};
