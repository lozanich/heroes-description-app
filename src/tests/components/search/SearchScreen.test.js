import React from "react";
import "@testing-library/jest-dom";
import { mount, shallow } from "enzyme";
import { MemoryRouter, Route } from "react-router";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe("Pruebas en el componente <SearchScreen />", () => {
  test("Debe de mostrarse correctamente con valores por defecto ", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen}></Route>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe("Busca un Heroe");
  });

  test("Debe de mostar a batman y el input con el query string", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchScreen}></Route>
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar un error si no se encuentra el Hero", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=goku"]}>
        <Route path="/search" component={SearchScreen}></Route>
      </MemoryRouter>
    );

    expect(wrapper.find(".alert-danger").exists()).toBe(true);
    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      "No se encontraron resultados"
    );
  });

  test("Debe de llamar el push del history", () => {
    const history = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=goku"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        ></Route>
      </MemoryRouter>
    );

    wrapper.find("input").simulate("change", {
      target: {
        name: "nameHero",
        value: "batman",
      },
    });

    wrapper.find("form").prop("onSubmit")({
      preventDefault() {},
    });

    expect(history.push).toHaveBeenCalledWith("?q=batman");
  });
});
