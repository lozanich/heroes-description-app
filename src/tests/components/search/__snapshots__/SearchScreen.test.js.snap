// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <SearchScreen /> Debe de mostar a batman y el input con el query string 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="form-group"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="nameHero"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search....
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card animate__animated animate__jello"
        style={
          Object {
            "maxWidth": 500,
            "minHeight": 840,
            "width": "18rem",
          }
        }
      >
        <img
          alt="Batman"
          className="card-img-top"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            Batman
          </h5>
          <ul
            className="list-group list-group-flush"
          >
            <li
              className="list-group-item"
            >
              DC Comics
            </li>
            <li
              className="list-group-item"
            >
              Detective Comics #27
            </li>
          </ul>
          <a
            className="btn btn-primary"
            href="/hero/dc-batman"
            onClick={[Function]}
          >
            Detalles
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en el componente <SearchScreen /> Debe de mostrarse correctamente con valores por defecto  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="form-group"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="nameHero"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search....
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Busca un Heroe
      </div>
    </div>
  </div>
</div>
`;
